### ISSUE TITLE
Training Recipes and Config System (YAML/JSON) for Reproducibility

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#283]
### User Story

> As a researcher, I want to define my entire training pipeline—including the model, dataset, optimizer, and hyperparameters—in a single YAML file, so that I can easily run, reproduce, and share my experiments without changing any code.

---

### **Phase 1: Configuration and Factories**

**Goal:** Define the C# classes that map to the YAML structure and create factories to instantiate objects from them.

#### **AC 1.1: Add YAML Dependency (1 point)**
**Requirement:** Add the necessary NuGet package for YAML deserialization.

-   [ ] In the `AiDotNet.csproj` file, add a `PackageReference` for `YamlDotNet` version 13.0.0 or higher.

#### **AC 1.2: Define Configuration POCOs (3 points)**
**Requirement:** Create the C# classes that will hold the configuration values.

-   [ ] Create a new folder: `src/Training/Configuration/`.
-   [ ] **`TrainingConfig.cs`:** Create a root class to hold all other configurations. It should have properties like `public ModelConfig Model { get; set; }`, `public DatasetConfig Dataset { get; set; }`, etc.
-   [ ] **`ModelConfig.cs`:** Create a class with properties `public string Name { get; set; }` and `public Dictionary<string, object> Params { get; set; }`.
-   [ ] **`DatasetConfig.cs`:** Create a class with properties `public string Name { get; set; }`, `public string Path { get; set; }`, and `public int BatchSize { get; set; }`.
-   [ ] **`OptimizerConfig.cs`:** Create a class with properties `public string Name { get; set; }` and `public double LearningRate { get; set; }`.
-   [ ] **`TrainerSettings.cs`:** Create a class with properties `public int Epochs { get; set; }` and `public bool EnableLogging { get; set; }`.

#### **AC 1.3: Implement Component Factories (5 points)**
**Requirement:** Create factory classes to build objects based on the configuration.

-   [ ] Create a new folder: `src/Training/Factories/`.
-   [ ] **`ModelFactory.cs`:**
    -   [ ] Create a `public static class ModelFactory`.
    -   [ ] Implement a method `public static IModel<T> Create(ModelConfig config)`.
    -   [ ] Inside, use a `switch` statement on `config.Name` to instantiate and return the correct model class (e.g., `case "SimpleCNN": return new SimpleCNN(config.Params);`).
-   [ ] **`OptimizerFactory.cs`:**
    -   [ ] Create a `public static class OptimizerFactory`.
    -   [ ] Implement `public static IOptimizer<T> Create(OptimizerConfig config, IModel<T> model)`.
    -   [ ] Use a `switch` statement on `config.Name` to return the correct optimizer (e.g., `case "Adam": return new Adam(model.GetParameters(), config.LearningRate);`).
-   [ ] Implement similar `DatasetFactory.cs` and `LossFunctionFactory.cs`.

---

### **Phase 2: The `Trainer` Harness**

**Goal:** Create the main `Trainer` class that uses the configuration and factories to run the training loop.

#### **AC 2.1: Implement the `Trainer` Class (8 points)**
**Requirement:** Build the class that orchestrates the entire training process.

-   [ ] Create a new folder: `src/Training/`.
-   [ ] Create a new file: `src/Training/Trainer.cs`.
-   [ ] Define a `public class Trainer<T>`.
-   [ ] **Constructor:**
    -   [ ] The constructor must accept a `TrainingConfig` object.
    -   [ ] Inside the constructor, use the factories from Phase 1 to instantiate the `_model`, `_dataset`, `_optimizer`, and `_lossFunction` and store them as private fields.
-   [ ] **`Run()` Method:**
    -   [ ] Implement a `public void Run()` method.
    -   [ ] Inside, implement a standard training loop: `for (int epoch = 0; epoch < _config.TrainerSettings.Epochs; epoch++)`.
    -   [ ] The loop should iterate through all batches from the `_dataset`.
    -   [ ] For each batch, it must perform the full training step: forward pass, loss calculation, backward pass, and optimizer step.
    -   [ ] If logging is enabled in the config, print the epoch number and average loss to the console.

#### **AC 2.2: Create Example YAML and Integration Test (5 points)**
**Requirement:** Provide a working example and a test to validate the end-to-end process.

-   [ ] Create a new folder and file: `examples/configs/simple-training.yaml`.
-   [ ] Populate the YAML file with a complete configuration for a simple model and dataset.
-   [ ] **Integration Test:**
    -   [ ] Create a new test file in the `tests` project.
    -   [ ] The test must first read and deserialize the `simple-training.yaml` file into a `TrainingConfig` object using `YamlDotNet`.
    -   [ ] Instantiate the `Trainer` with the loaded config object.
    -   [ ] Call `Trainer.Run()` for a small number of epochs (e.g., 2).
    -   [ ] **Assert** that the method completes without throwing an exception.
    -   [ ] (Optional but recommended) Assert that the model's weights have changed from their initial values.

---

### **Definition of Done**

-   [ ] All checklist items are complete.
-   [ ] A user can define a complete training job in a single YAML file.
-   [ ] The `Trainer` class can successfully parse the config, build the required components, and run the training loop.
-   [ ] The integration test proves the end-to-end functionality.
-   [ ] All new code meets the project's >= 90% test coverage requirement.
---

## ⚠️ CRITICAL ARCHITECTURAL REQUIREMENTS

**Before implementing this user story, you MUST review:**
- **📋 Full Requirements:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
- **📐 Project Rules:** [`.github/PROJECT_RULES.md`](../.github/PROJECT_RULES.md)

#### 1. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected static readonly INumericOperations<T> NumOps = MathHelper.GetNumericOperations<T>();` in base class
- [ ] NEVER hardcode `double`, `float`, or specific numeric types - use generic `T`
- [ ] NEVER use `default(T)` - use `NumOps.Zero` instead
- [ ] Use `NumOps.Zero`, `NumOps.One`, `NumOps.FromDouble()` for values
- [ ] Use `NumOps.Add()`, `NumOps.Multiply()`, etc. for arithmetic
- [ ] Use `NumOps.LessThan()`, `NumOps.GreaterThan()`, etc. for comparisons

#### 2. Inheritance Pattern (REQUIRED)
- [ ] Create `I{FeatureName}.cs` in `src/Interfaces/` (root level, NOT subfolders)
- [ ] Create `{FeatureName}Base.cs` in `src/{FeatureArea}/` inheriting from interface
- [ ] Create concrete classes inheriting from Base class (NOT directly from interface)

#### 3. PredictionModelBuilder Integration (REQUIRED)
- [ ] Add private field: `private I{FeatureName}<T>? _{featureName};` to `PredictionModelBuilder.cs`
- [ ] Add Configure method taking ONLY interface (no parameters):
  ```csharp
  public IPredictionModelBuilder<T, TInput, TOutput> Configure{FeatureName}(I{FeatureName}<T> {featureName})
  {
      _{featureName} = {featureName};
      return this;
  }
  ```
- [ ] Use feature in `Build()` with default: `var {featureName} = _{featureName} ?? new Default{FeatureName}<T>();`
- [ ] Verify feature is ACTUALLY USED in execution flow

#### 4. Beginner-Friendly Defaults (REQUIRED)
- [ ] Constructor parameters with defaults from research/industry standards
- [ ] Document WHY each default was chosen (cite papers/standards)
- [ ] Validate parameters and throw `ArgumentException` for invalid values

#### 5. Property Initialization (CRITICAL)
- [ ] NEVER use `default!` operator
- [ ] String properties: `= string.Empty;`
- [ ] Collections: `= new List<T>();` or `= new Vector<T>(0);`
- [ ] Numeric properties: appropriate default or `NumOps.Zero`

#### 6. Class Organization (REQUIRED)
- [ ] One class/enum/interface per file
- [ ] ALL interfaces in `src/Interfaces/` (root level)
- [ ] Namespace mirrors folder structure (e.g., `src/Regularization/` → `namespace AiDotNet.Regularization`)

#### 7. Documentation (REQUIRED)
- [ ] XML documentation for all public members
- [ ] `<b>For Beginners:</b>` sections with analogies and examples
- [ ] Document all `<param>`, `<returns>`, `<exception>` tags
- [ ] Explain default value choices

#### 8. Testing (REQUIRED)
- [ ] Minimum 80% code coverage
- [ ] Test with multiple numeric types (double, float)
- [ ] Test default values are applied correctly
- [ ] Test edge cases and exceptions
- [ ] Integration tests for PredictionModelBuilder usage

---

**⚠️ Failure to follow these requirements will result in repeated implementation mistakes and PR rejections.**

**See full details:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)

**Comment by ooples** [comment]
Agent: Agent-Recipes
Branch: feature/issue-283-training-recipes-and-config-system-yaml-json-for-reproducibility
Worktree: .worktrees\feature-issue-283-training-recipes-and-config-system-yaml-json-for-reproducibility

User Story
As a .NET ML engineer, I want to deliver "Training Recipes and Config System (YAML/JSON) for Reproducibility" so that AiDotNet users can rely on a robust, reproducible, and .NET Framework 4.6–compatible implementation.

Acceptance Criteria
- Design reviewed and linked in PR description
- API surface aligned with existing patterns in src and documented in README changes
- Unit tests added (xUnit) covering new logic with >= 90% line/branch coverage for changed files
- Builds succeed for 
et8.0 and 
et462 on CI and locally
- Benchmarks or microperf notes included where performance-sensitive (if applicable)
- Examples or docs updated where applicable

Plan of Action
- [ ] Confirm scope and traceability to this issue
- [ ] Design draft and API sketch (commit as docs or comment)
- [ ] Implement feature behind tests
- [ ] Add tests (xUnit) and ensure coverage >= 90%
- [ ] Run dotnet build + dotnet test --collect:"XPlat Code Coverage" for net8.0 and net462
- [ ] Update docs (README, samples) if needed
- [ ] Open PR linking this issue ("Fixes #283")

Definition of Done
- All acceptance criteria met
- Build green on CI
- Code coverage >= 90% for new/changed code
- Maintains compatibility with .NET Framework 4.6 (
et462)
